crates/storage/provider/src/post_state.rs

```diff
@@ -278,52 +177,87 @@ impl PostState {
         calculate_receipt_root_ref(self.receipts().iter().map(Into::into))
     }
 
-    /// Get the number of transitions causing this [PostState]
-    pub fn transitions_count(&self) -> TransitionId {
-        self.current_transition_id
-    }
-
+    // todo: note overwrite behavior, i.e. changes in `other` take precedent
     /// Extend this [PostState] with the changes in another [PostState].
-    pub fn extend(&mut self, other: PostState) {
-        if other.changes.is_empty() {
-            return
+    pub fn extend(&mut self, mut other: PostState) {
+        // Update plain state
+        self.accounts.extend(other.accounts);
+        for (address, their_storage) in other.storage {
+            let our_storage = self.storage.entry(address).or_default();
+            if their_storage.wiped {
+                our_storage.wiped = true;
+                our_storage.storage.clear();
+            }
+            our_storage.storage.extend(their_storage.storage);
         }
 
-        self.changes.reserve(other.changes.len());
+        // Insert account change sets
+        for (block_number, account_changes) in std::mem::take(&mut other.account_changes) {
+            let block = self.account_changes.entry(block_number).or_default();
+            for (address, account) in account_changes {
+                // TODO: Note about why we skip accounts we already have
+                if block.contains_key(&address) {
+                    continue
+                }
+                block.insert(address, account);
+            }
+        }
+
+        // Insert storage change sets
+        for (block_number, storage_changes) in std::mem::take(&mut other.storage_changes) {
+            for (address, their_storage) in storage_changes {
+                let our_storage = self
+                    .storage_changes
+                    .entry(block_number)
+                    .or_default()
+                    .entry(address)
+                    .or_default();
+
+                if their_storage.wiped {
+                    our_storage.wiped = true;
+                }
```

Comment:
nit - we could move this back into an `add_and_apply` function, but no big deal to have it inlined

Reply:
I would leave it as it is for now, as to not introduce conflict, but can be changed later.